src/MergeContext.js

```diff
@@ -198,19 +216,269 @@ class StatusChecks
+//   c) with a stale staged commit which will be deleted before creation a new one
+// staged: with a staged commit that has not been merged into the base branch
+// merged: with a staged commit that has been merged into the base branch
+// Here, PR "staged commit" is a commit at the tip of the staging branch
+// pointed to by the PR tag. If a PR tag does not exist or does not point to
+// the tip of the staging branch, then the PR does not have a staged commit.
+class PrState
+{
+    // treat as private; use static methods below instead
+    constructor(state) {
+        this._state = state;
+    }
+
+    static Brewing() { return new PrState(-1); }
+    static Staged() { return new PrState(0); }
+    static Merged() { return new PrState(1); }
+
+    brewing() { return this._state < 0; }
+    staged() { return this._state === 0; }
+    merged() { return this._state > 0; }
+
+    toString() {
+        if (this.brewing())
+            return "brewing";
+        if (this.staged())
+            return "staged";
+        assert(this.merged());
+        return "merged";
+    }
+}
+
+// external limits for PullRequest::process() freedoms
+class PrRestrictions
+{
+    constructor() {
+        this._banStaging = false; // do not stage; only valid for brewing PRs
+        this._banInstantRetries = false; // do not ask the caller to re-process() immediately
+    }
+
+    stagingBanned() {
+        return this._banStaging;
+    }
+
+    instantRetriesBanned() {
+        return this._banInstantRetries;
+    }
+
+    banStaging(bool) {
+        this._banStaging = bool;
+        return this;
+    }
+
+    banInstantRetries(bool) {
+        this._banInstantRetries = bool;
+        return this;
+    }
+}
+
+// Determines the feature branch position relative to its base branch: ahead,
+// merged, or diverged. This class essentially decides whether/which one of
+// the two branches is fully contained in another.
+class BranchPosition
+{
+    constructor(baseRef, featureRef) {
+        assert(baseRef !== featureRef);
+        this._baseRef = baseRef;
+        this._featureRef = featureRef;
+        this._status = null;
+    }
+
+    async compute() {
+        this._status = await GH.compareCommits(this._baseRef, this._featureRef);
+    }
+
+    // feature > base:
+    // the feature branch contains the base branch and some additional commits
+    ahead() {
+        assert(this._status);
+        return this._status === "ahead";
+    }
+
+    // feature <= base:
+    // either both branches point to same commit (e.g., just merged) or
+    // there are new commits on base since the feature branch was merged
+    merged() {
+        assert(this._status);
+        return this._status === "identical" || this._status === "behind";
+    }
+
+    // neither ahead nor merged:
+    // both base and the feature branch have unique-to-them commits
+    diverged() {
+        assert(this._status);
+        return this._status === "diverged";
```

Comment:
To better match our method description and to be on the safe side in case GitHub adds more statuses:

```suggestion
         // this._status should be "diverged", but we also handle any new/unexpected statuses here
        return !(this.ahead() || this.merged());
```

N.B. I am curious, what status will GitHub report if the commits have no common ancestor at all? I understand that this is highly unlikely in a mature repository, but I think it is technically possible if somebody force-pushes a completely different/new master branch after the feature branch was branched off the old master. Would that result in a "diverged" `_status`? The change suggested above would make the answer irrelevant.